addons/hr_expense/migrations/10.0.2.0/post-migration.py

```diff
@@ -1,9 +1,69 @@
 # -*- coding: utf-8 -*-
 # Copyright 2017 Le Filament (<https://le-filament.com>)
+# Copyright 2017 Tecnativa - Pedro M. Baeza
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
 from openupgradelib import openupgrade
 
 
+def set_expense_sheet_address(env):
+    """Set address for each employee."""
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE hr_expense_sheet s
+        SET address_id = e.address_home_id
+        FROM hr_employee e
+        WHERE s.employee_id = e.id""",
+    )
+
+
+def set_expense_responsible(env):
+    """Set responsible on expenses by setting them to department manager user.
+    """
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE hr_expense_sheet s
+        SET responsible_id = r.user_id
+        FROM hr_department d,
+             hr_employee e,
+             resource_resource r
+        WHERE s.department_id = d.id
+            AND d.manager_id = e.id
+            AND r.id = e.resource_id""",
+    )
+
+
+def set_expense_states(env):
+    """Set correct states on hr.expense"""
+    openupgrade.map_values(
+        env.cr, openupgrade.get_legacy_name('state'), 'state', [
+            ('submit', 'reported'),
+            ('approved', 'reported'),
+            ('post', 'done'),
+            ('cancel', 'refused'),
+        ], table='hr_expense',
+    )
+
+
+def fill_expense_product_account(env):
+    """Set correct account corresponding to product subject to this expense."""
+    for expense in env['hr.expense'].search([]):
+        expense.account_id = expense.product_id.with_context(
+            force_company=expense.company_id.id,
+        ).product_tmpl_id._get_product_accounts()['expense']
```

Comment:
You could build a map of account values for all unique combinations of (product, company) extracted with an sql query

```SELECT product_id, company_id, ARRAY_AGG(id) FROM hr_expense group by product_id, company_id```

and then write the expenses in groups. If you have a large amount of expenses it could improve performance. But it's just a suggestion, I'll leave it up to you.

Reply:
It's not so easy, Stefan, due to:

- The account can be at product or category level.
- Properties themselves have 3 possible level of values: general default value, company default value, record value.

So this performance trade-off avoids possible problems.

Reply:
Pedro, if you retrieve the account for every combination of (product, company) you have all of that covered.

Reply:
It still lacks the default value without company for those without specific value

Reply:
Let me clarify: 
```
self.env.cr.execute("SELECT product_id, company_id, ARRAY_AGG(id) FROM hr_expense group by product_id company_id")
for product_id, company_id, expense_ids in self.env.cr.fetchall():
    self.env['hr.expense'].browse(expense_ids).write({
        'account_id': self.env['product.product'].with_context(
            force_company=company_id).browse(
                product_id).product_tmpl_id._get_product_accounts()['expense'].id})
```
I don't think you will run into problems of missing accounts with that, do you?